dvc/remote/base.py

```diff
@@ -505,17 +507,37 @@ def pull(self, cache, named_cache, jobs=None, show_checksums=False):
 
         return ret
 
-    def transfer(self, from_fs, from_info, jobs=None, no_progress_bar=False):
-        from dvc.objects.transfer import transfer
-
-        return transfer(
+    def transfer(
+        self,
+        from_fs,
+        from_info,
+        jobs=None,
+        update=False,
+        no_progress_bar=False,
+    ):
+        # When running import-url --to-remote / add --to-remote/-o ... we
+        # assume that it is unlikely that the odb will contain majority of the
+        # hashes, so we transfer everything as is (even if that file might
+        # already be in the cache) and don't waste an upload to scan the layout
+        # of the source location. But when doing update --to-remote, there is
+        # a high probability that the odb might contain some of the hashes, so
+        # we first calculate all the hashes (but don't transfer anything) and
+        # then only update the missing cache files.
+
+        upload = not (update and from_fs.isdir(from_info))
```

Comment:
Why do we have to upload if we are updating a file though?

Reply:
because there is no point of calculating the hash, since we'd end up with downloading one time (calculating the hash) + odb.add()ing it (since it is unlikely that the file's hash won't be unchanged when you do update) which would cost download + upload. This optimization only works for directories

Reply:
@isidentical Makes sense :+1: Btw, looks like if we would've saved to state during import, then we would be able to quickly tell that this file indeed changed and save us that downloading(for calculating the hash) that you've mentioned. Just a thought.